Readme.md
[![GitHub tag](https://img.shields.io/github/tag/camel-tooling/vscode-wsdl2rest.svg?style=plastic)]()
[![CircleCI](https://circleci.com/gh/circleci/circleci-docs.svg?style=shield)](https://circleci.com/gh/camel-tooling/vscode-wsdl2rest)
[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=vscode-wsdl2rest&metric=alert_status)](https://sonarcloud.io/dashboard?id=vscode-wsdl2rest)
[![License](https://img.shields.io/badge/license-Apache%202-blue.svg)]()
[![Gitter](https://img.shields.io/gitter/room/camel-tooling/Lobby.js.svg)](https://gitter.im/camel-tooling/Lobby)

This VS Code extension is now deprecated.

# WSDL to Camel Rest DSL for Visual Studio Code
This preview release of the extension adds WSDL to REST support to [Visual Studio Code](https://code.visualstudio.com/).

 By specifying an existing WSDL file, you can use this extension to generate a Camel Rest DSL + CXF solution for REST-style access. The WSDL file can be located either on your local file system or from an accessible web URL.

Note: For this preview release, the **WSDL to Camel Rest DSL** extension creates Spring, Spring Boot, or Blueprint configurations. Java configurations are not supported yet.

## Contact Us
If you run into an issue or have a suggestion, you can report it by [creating a new issue on GitHub](https://github.com/camel-tooling/vscode-wsdl2rest/issues).

## How to install
You can download this **WSDL to Camel Rest DSL** extension from the Visual Studio Code Marketplace on its own or as part of the [Extension pack for Apache Camel by Red Hat](https://marketplace.visualstudio.com/items?itemName=camel-tooling.apache-camel-extension-pack or as its own) [wsdl2rest extension by Red Hat](https://marketplace.visualstudio.com/items?itemName=camel-tooling.vscode-wsdl2rest).

After you install VS Code, follow these steps:
1. In VS Code, select **View > Extensions**.
2. Search for **Camel**.
3. Select the **Extension Pack for Apache Camel** or the **WSDL to Camel Rest DSL** option and then click **Install**.

## Before you begin
Before you use this **WSDL to Camel Rest DSL** extension, you should create a mavenized Camel- or Fuse-based project, such as one of the following:

* A project that you create with the [Project Initializer by Red Hat VS Code extension](https://marketplace.visualstudio.com/items?itemName=redhat.project-initializer).   
Note: This extension is included in the [Extension pack for Apache Camel](https://marketplace.visualstudio.com/items?itemName=camel-tooling.apache-camel-extension-pack).

* A standalone Camel- or Fuse-based Spring project based on an existing example or a quickstart, for example:  
https://github.com/apache/camel/tree/master/examples/camel-example-spring  
or  
https://github.com/jboss-fuse/quickstarts/tree/master/beginner/camel-cbr

* A Maven project created by using [mvn:archetype](http://camel.apache.org/camel-maven-archetypes.html) with camel-archetype-spring, camel-archetype-blueprint, or camel-archetype-spring-boot.

After you create a project, you can then use the **WSDL to Camel Rest DSL** extension to generate the Camel Rest DSL + CXF solution for REST-style access. 

## Using the extension
To generate your Camel configuration from a WSDL file:

1. In VS Code, create a new workspace and then add the Camel- or Fuse-based project folder to the workspace.

2. Select **View > Command Palette** and then type **wsdl2rest**.  

      ![wsdl2rest in Command Palette](./images/wsdl2rest-dropdown.png "wsdl2rest in Command Palette")

3. Select whether to use a local WSDL file or a WSDL URL.

  * For a local WSDL file: In the File dialog that appears, browse to find the WSDL file in your local file system.
  
  * For a WSDL file URL: In the drop-down that appears, type the URL to the WSDL that you want to access.

4. Specify the DSL (Spring or Blueprint) for the generated Camel configuration.  

5. Specify the output directory for the generated CXF artifacts. The default directory is `src/main/java`.  
  
      **Important note:** You should change the output directory to a different folder if you do not want the extension to overwrite  existing classes with the same package or class name.

6. (Optional) Specify the address for the running JAX-WS endpoint. The default address is http://localhost:8080/somepath  
  
      **Note:** If you accept the default address and the WSDL file's `service/port` binding specifies a SOAP address location, then the generated Camel configuration uses that SOAP address instead of the default address.
  
7. (Optional) Specify the address for the generated JAX-RS endpoint. The default address is http://localhost:8081/jaxrs  

      The extension generates:
  * A Camel Rest configuration in your chosen DSL
  * The CXF artifacts to harness the power of your SOAP-based JAX-WS service in a RESTful way


8. Before you build or run the generated code, check the Maven configuration for your project and update it to include the following dependencies as needed:

### Spring- and Spring Boot-based projects

```xml
    <!-- wsdl2rest dependencies -->
    <dependency>
        <groupId>org.jboss.spec.javax.ws.rs</groupId>
        <artifactId>jboss-jaxrs-api_2.0_spec</artifactId>
        <version>1.0.0.Final-redhat-1</version>
    </dependency>
    <dependency>
        <groupId>org.apache.camel</groupId>
        <artifactId>camel-jackson</artifactId>
        <version>${camel.version}</version>
    </dependency>
    <dependency>
        <groupId>org.apache.camel</groupId>
        <artifactId>camel-cxf</artifactId>
        <version>${camel.version}</version>
    </dependency>
    <dependency>
        <groupId>org.apache.camel</groupId>
        <artifactId>camel-jetty</artifactId>
        <version>${camel.version}</version>
    </dependency>
```

### Blueprint-based projects

```xml
    <!-- wsdl2rest dependencies -->
    <dependency>
        <groupId>org.jboss.spec.javax.ws.rs</groupId>
        <artifactId>jboss-jaxrs-api_2.0_spec</artifactId>
        <version>1.0.0.Final-redhat-1</version>
    </dependency>
    <dependency>
        <groupId>org.apache.camel</groupId>
        <artifactId>camel-jackson</artifactId>
        <version>${camel.version}</version>
    </dependency>
    <dependency>
        <groupId>org.apache.camel</groupId>
        <artifactId>camel-cxf</artifactId>
        <version>${camel.version}</version>
    </dependency>
    <dependency>
        <groupId>org.apache.camel</groupId>
        <artifactId>camel-servlet</artifactId>
        <version>${camel.version}</version>
    </dependency>
```



CHANGELOG.md
# Change Log

All notable changes to the "vscode-wsdl2rest" extension will be documented in this file.

## 0.0.16

## 0.0.15

- Upgrade internal use of Wsdl2rest from 7.9.0 version to 7.10.0 version

## 0.0.14

- This VS Code extension is now deprecated.

## 0.0.13

- Upgrade internal use of Wsdl2rest from 7.7.0 version to 7.9.0 version

## 0.0.12

- Upgrade internal use of Wsdl2rest from 7.5.0 version to 7.7.0 version

## 0.0.11

- Upgrade internal use of Wsdl2rest from Early Access 7.3.0 to Publicly released 7.5.0 version

## 0.0.10

- Update to naming approved by Red Hat legal

## 0.0.9

- move to Red Hat publisher to gain visibility

## 0.0.8

- updating version of wsdl2rest to 0.8.0.fuse-730094-redhat-00001 (final version for Fuse 7.3 release)
- updated spring-boot-maven-plugin version to 2.1.4.RELEASE (April 2019 release)
- updating readme upstream for published vscode-wsdl2rest release

## 0.0.7

- updated readme with additional help and clarification
- updated version of wsdl2rest to 0.8.0.fuse-73090

## 0.0.6

- updated to use wsdl2rest version 0.8.0.fuse-730050
- made jax-ws and jax-rs URL prompts optional to use the underlying wsdl2rest ability to pull the jax-ws URL directly from the WSDL if it's specified there and using the standard wsdl2rest defaults rather than overwriting them each time https://github.com/camel-tooling/vscode-wsdl2rest/issues/34

## 0.0.5

- addressed issue with URL-based WSDL files https://github.com/camel-tooling/vscode-wsdl2rest/issues/36 
- addressed issue with focus loss https://github.com/camel-tooling/vscode-wsdl2rest/issues/35

## 0.0.4

- addressed issue https://github.com/camel-tooling/vscode-wsdl2rest/issues/30 

## 0.0.3

- Initial release
- Support creation of Camel Rest DSL artifacts from local WSDL file
- Support creation of Camel Rest DSL artifacts from remote (URL-accessible) WSDL file
- Published to VS Code Marketplace


src/extension.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

import * as child_process from 'child_process';
import * as fs from 'fs-extra';
import * as path from 'path';
import * as requirements from './requirements';
import * as utils from './utils';
import * as vscode from 'vscode';
import fileUrl from 'file-url';

let outputChannel: vscode.OutputChannel;
let wsdl2restProcess: child_process.ChildProcess;
let javaExecutablePath: string;
let wsdl2restExecutablePath: string;
let resourcesPath: string;
let wsdlFileUri: string;
let outputDirectory: string;
let dsl: string;
let jaxrs: string;
let jaxws: string;
let readmePath: string;
let storagePath : string | undefined;

export function activate(context: vscode.ExtensionContext) {
	
	wsdl2restExecutablePath = context.asAbsolutePath(path.join('./', 'jars/','wsdl2rest.jar'));
	resourcesPath = context.asAbsolutePath(path.join('./', 'resources/'));
	outputChannel = vscode.window.createOutputChannel("WSDL2Rest");
	context.subscriptions.push(vscode.commands.registerCommand('extension.wsdl2rest.local', () => callWsdl2RestViaUIAsync(false)));
	context.subscriptions.push(vscode.commands.registerCommand('extension.wsdl2rest.url', () => callWsdl2RestViaUIAsync(true)));
}

// note this is just exposed for testing purposes
export function getStoragePath() : string {
	return storagePath;
}

function callWsdl2RestViaUIAsync(useUrl: boolean): Promise<void> {
	return new Promise <void> ( async (resolve, reject) => {
		await askForUserInputs(useUrl)
		.then( () => {
			callWsdl2Rest(wsdl2restExecutablePath)
				.then( success => {
					if (!success) {
						vscode.window.showErrorMessage("Unable to create the Wsdl2Rest files.");
						reject();
					}
					resolve();
					return success;
				})
				.catch(err => {
					console.error("Wsdl2Rest execution return code: " + err);
					reject();
					return err;
				});
		})
		.catch(err => {
			console.error("Error retrieving the required user inputs. " + err);
			reject();
			return err;
		});
	});
}

function validateEndpointUrl (text: string): string {
	if (!utils.isAValidUrl(text)) {
		return "WSDL URL not valid.";
	}
	return null;
}

function validateWsdlUrl (text: string): string {
	if (!utils.isAValidUrl(text)) {
		return "WSDL URL not valid.";
	}
	if (!utils.isFileAvailable(text)) {
		return "WSDL file not available at URL specified.";
	}
	return null;
}

function validatePath (text: string): string {
	if (!utils.isValidPath(text)) {
		return "Invalid output folder specified.";
	}
	return null;
}

function askForUserInputs(useUrl: boolean): Promise<void> {
	return new Promise( async (resolve, reject) => {
		try {
			let fileUri;
			if (!useUrl) {
				fileUri = await vscode.window.showOpenDialog(utils.Options);
			} else {
				fileUri = await vscode.window.showInputBox({
					prompt: 'WSDL URL',
					placeHolder: 'Provide the URL for the WSDL file',
					ignoreFocusOut: true,
					validateInput: (text: string) => validateWsdlUrl(text)
				});
			}
			if (fileUri === undefined) {
				return reject("No WSDL specified.");
			}
			if (fileUri && Array.isArray(fileUri)) {
				wsdlFileUri = fileUri[0] + "";
			} else if (fileUri) {
				wsdlFileUri = fileUri;
			} else {
				return reject("WSDL not valid.");
			}
			utils.printDebug("WSDL File URI: " + wsdlFileUri);

			dsl = await vscode.window.showQuickPick(
				[
					utils.DslType.Spring, 
					utils.DslType.Blueprint
				],
				{
					placeHolder: 'Specify which DSL to generate the Camel configuration for'
				}
			);
			if (!dsl || dsl === undefined) {
				return reject("No valid DSL Type selected.");
			}
			utils.printDebug("DSL Type: " + dsl);

			outputDirectory = await vscode.window.showInputBox({
				prompt: 'Output Directory',
				placeHolder: 'Enter the output directory for generated artifacts',
				ignoreFocusOut: true,
				value: 'src/main/java',
				validateInput: (text: string) => validatePath(text)
			});
			if (!outputDirectory || outputDirectory === undefined) {
				return reject("No valid output folder specified.");
			}
			utils.printDebug("Ouput Folder: " + outputDirectory);

			jaxws = await vscode.window.showInputBox({
				prompt: 'JAXWS Endpoint',
				placeHolder: 'Enter the address for the running jaxws endpoint (defaults to http://localhost:8080/somepath)',
				ignoreFocusOut: true,
				validateInput: (text: string) => validateEndpointUrl(text)
			});
			if(jaxws === undefined) {
				return reject("Action aborted at JAXWS Endpoint input box step.");
			}
			utils.printDebug("JAXWS Endpoint: " + jaxws);

			jaxrs = await vscode.window.showInputBox({
				prompt: 'JAXRS Endpoint',
				placeHolder: 'Enter the address for the jaxrs endpoint (defaults to http://localhost:8081/jaxrs)',
				ignoreFocusOut: true,
				validateInput: (text: string) => validateEndpointUrl(text)
			});
			if(jaxrs === undefined){
				return reject("Action aborted at JAXRS Endpoint input box step.");
			}
			utils.printDebug("JAXRS Endpoint: " + jaxrs);
			return resolve();
		} catch (error) {
			console.error(error);
			return reject(error);
		}
	});
}

function callWsdl2Rest(wsdl2restExecutablePath: string): Promise<boolean> {
	return new Promise( (resolve, reject) => {
		try {
			if (vscode.workspace.workspaceFolders && vscode.workspace.workspaceFolders[0]) {
				storagePath = vscode.workspace.workspaceFolders[0].uri.fsPath;
			}
			if (!storagePath) { // i.e. the workspace was empty
				storagePath = utils.getTempWorkspace();
			}

			if (outputDirectory.endsWith('/java')) {
				outputDirectory = outputDirectory.substring(0, outputDirectory.indexOf('/java'));
			}

			let outPath: string = path.join(storagePath, outputDirectory);
			
			if (!(wsdlFileUri.startsWith('http:') || wsdlFileUri.startsWith('https:'))) {
				if (!wsdlFileUri.startsWith('file:')) {
					wsdlFileUri = fileUrl(wsdlFileUri);
				}
			}
			
			if (!fs.existsSync(outPath)) {
				vscode.window.showInformationMessage(`Creating Wsdl2Rest Java output directory: ` + outPath);
				fs.ensureDirSync(outPath);
			}
			
			var restContextPath;
			var rawContextPath: any;

			const isBlueprint: boolean = dsl === utils.DslType.Blueprint;
			const isSpringBoot: boolean = dsl === utils.DslType.SpringBoot;
			const isSpring: boolean = dsl === utils.DslType.Spring;

			if (isBlueprint) {
				rawContextPath = outputDirectory + '/resources/OSGI-INF/blueprint/blueprint.xml';
			} else if (isSpringBoot) {
				rawContextPath = outputDirectory + '/resources/camel-context.xml';
			} else if (isSpring) {
				rawContextPath = outputDirectory + '/resources/META-INF/spring/camel-context.xml';
			}
			restContextPath = path.join(storagePath, rawContextPath);

			let readmeRoot = outputDirectory.substring(0, outputDirectory.indexOf('/'));
			let newReadmePath: any;
			if (/src/.test(readmeRoot)) {
				newReadmePath = path.join(storagePath, 'wsdl2rest.readme.md');
			} else {
				newReadmePath = path.join(storagePath, readmeRoot, 'wsdl2rest.readme.md');
			}

			if (outputChannel) {
				outputChannel.clear();
				outputChannel.show();
			}

			requirements.resolveRequirements()
				.then(requirements => {
					let originalLog4JProps = resourcesPath.substring(0, resourcesPath.lastIndexOf(path.sep)+1) + "log4j.properties";
					let newLogsFolder = storagePath + path.sep + 'config';
					let newLog4JProps = newLogsFolder + path.sep + 'logging.properties';
					fs.copySync(path.resolve(originalLog4JProps), newLog4JProps);
					utils.printDebug("New config folder: " + newLogsFolder);

					let log4jConfigPath: string = fileUrl(newLog4JProps);
					utils.printDebug("Log4J Config: " + log4jConfigPath);
					javaExecutablePath = path.resolve(requirements.java_home + '/bin/java');

					let readmePath = resourcesPath.substring(0, resourcesPath.lastIndexOf(path.sep)+1) + 'wsdl2rest.readme.md';
					fs.copySync(path.resolve(readmePath), newReadmePath);
					utils.printDebug("New readme: " + newReadmePath);

					let contextType = "--camel-context";
					if (isBlueprint) {
						contextType = "--blueprint-context";
					}
					let args:string[];
					args = [ "-Dlog4j.configuration=" + log4jConfigPath, 
						"-jar", wsdl2restExecutablePath,
						"--wsdl", wsdlFileUri,
						"--out", outPath,
						contextType, restContextPath];
					if (!utils.isEmpty(jaxrs)) {
						args.push("--jaxrs");
						args.push(jaxrs);
					}
					if (!utils.isEmpty(jaxws)) {
						args.push("--jaxws");
						args.push(jaxws);
					}

					utils.printDebug("Java Binary: " + javaExecutablePath);
					utils.printDebug("Wsdl2Rest JAR: " + wsdl2restExecutablePath);
					utils.printDebug("Java Call: " + javaExecutablePath + "\n\t" + args);
					outputChannel.append("Executing Wsdl2Rest...\n");
					wsdl2restProcess = child_process.spawn(javaExecutablePath, args);

					wsdl2restProcess.stdout.on('data', function (data) {
						outputChannel.append(`${data} \n`);
						utils.printDebug(data);
					});
					wsdl2restProcess.stderr.on('data', function (data) {
						outputChannel.append(`${data} \n`);
						utils.printDebug(data);
					});
					wsdl2restProcess.on("close", (code, signal) => {
						if (code === 0) {
							vscode.window.showInformationMessage('Created CXF artifacts for specified WSDL at ' + outputDirectory);
							vscode.window.showInformationMessage('Created ' + rawContextPath);
						} else {
							vscode.window.showErrorMessage(`Wsdl2Rest did not generate artifacts successfully - please check the output channel for details`);
						}
						outputChannel.append("\nProcess finished. Return code " + code + ".\n\n");
						resolve(code === 0);
					});				
				});
			} catch (error) {
				console.error(error);
				reject(error);
			}
		});
}
